qiskit/extensions/quantum_initializer/state_preparation.py

```diff
@@ -0,0 +1,462 @@
+                initialize_circuit.append(SGate(), [q[qubit]])
+            elif param == "l":  # |-i>
+                initialize_circuit.append(HGate(), [q[qubit]])
+                initialize_circuit.append(SdgGate(), [q[qubit]])
+
+        return initialize_circuit
+
+    def _define_from_int(self):
+        q = QuantumRegister(self.num_qubits, "q")
+        initialize_circuit = QuantumCircuit(q, name="init_def")
+
+        # Convert to int since QuantumCircuit converted to complex
+        # and make a bit string and reverse it
+        intstr = f"{int(np.real(self.params[0])):0{self.num_qubits}b}"[::-1]
+
+        # Raise if number of bits is greater than num_qubits
+        if len(intstr) > self.num_qubits:
+            raise QiskitError(
+                "StatePreparation integer has %s bits, but this exceeds the"
+                " number of qubits in the circuit, %s." % (len(intstr), self.num_qubits)
+            )
+
+        for qubit, bit in enumerate(intstr):
+            if bit == "1":
+                initialize_circuit.append(XGate(), [q[qubit]])
+
+        return initialize_circuit
+
+    def _default_synthesis(self):
+        """Calculate a subcircuit that implements this initialization
+
+        Implements a recursive initialization algorithm, including optimizations,
+        from "Synthesis of Quantum Logic Circuits" Shende, Bullock, Markov
+        https://arxiv.org/abs/quant-ph/0406176v5
+
+        Additionally implements some extra optimizations: remove zero rotations and
+        double cnots.
+        """
+        # call to generate the circuit that takes the desired vector to zero
+        disentangling_circuit = self.gates_to_uncompute()
+
+        # invert the circuit to create the desired vector from zero (assuming
+        # the qubits are in the zero state)
+        initialize_instr = disentangling_circuit.to_instruction().inverse()
+
+        q = QuantumRegister(self.num_qubits, "q")
+        initialize_circuit = QuantumCircuit(q, name="init_def")
+        initialize_circuit.append(initialize_instr, q[:])
+
+        return initialize_circuit
+
+    @staticmethod
+    def get_num_qubits(num_qubits, params):
+        if isinstance(params, str):
+            num_qubits = len(params)
+        elif isinstance(params, int):
+            if num_qubits is None:
+                num_qubits = int(math.log2(params)) + 1
+        else:
+            num_qubits = math.log2(len(params))
+
+            # Check if param is a power of 2
+            if num_qubits == 0 or not num_qubits.is_integer():
+                raise QiskitError("Desired statevector length not a positive power of 2.")
+
+            # Check if probabilities (amplitudes squared) sum to 1
+            if not math.isclose(sum(np.absolute(params) ** 2), 1.0, abs_tol=_EPS):
+                raise QiskitError("Sum of amplitudes-squared does not equal one.")
+
+            num_qubits = int(num_qubits)
+        return num_qubits
```

Comment:
the logic itself was just moved over from what was originally in the `Initialize` and it was made public in case we needed to call it in the `Initialize` class. But after refactoring actually i don't think we need to call it anymore in the `Initialize` class because we can just access `state_prep.num_qubits` so I can probably make this a private method again